File: activities/DisplayActivity.java
package com.chad.gads2022_java_kotlin.activities;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.chad.gads2022_java_kotlin.R;
import com.chad.gads2022_java_kotlin.adapters.DisplayAdapter;
import com.chad.gads2022_java_kotlin.app.Constants;
import com.chad.gads2022_java_kotlin.app.Util;
import com.chad.gads2022_java_kotlin.models.Repository;
import com.chad.gads2022_java_kotlin.models.SearchResponse;
import com.chad.gads2022_java_kotlin.retrofit.GithubAPIService;
import com.chad.gads2022_java_kotlin.retrofit.RetrofitClient;
import com.google.android.material.navigation.NavigationView;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.realm.Realm;
import io.realm.RealmResults;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;



public class DisplayActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

	private static final String TAG = DisplayActivity.class.getSimpleName();

	private DrawerLayout mDrawerLayout;
	private RecyclerView mRecyclerView;
	private DisplayAdapter mDisplayAdapter;
	private List<Repository> browsedRepositories;
	private GithubAPIService mService;
	private Realm mRealm;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_display);

		//Initialize
		initialize();

		//Get Bundle Data
		getBundleData();
	}

	//Initialize Views
	private void initialize() {

		//Initialize Toolbar and set title
		Toolbar toolbar = findViewById(R.id.toolbar_display);
		setSupportActionBar(toolbar);
		getSupportActionBar().setTitle("Showing Browsed Results");

		//Initialize RecyclerView and set LayoutManager
		mRecyclerView = findViewById(R.id.recyclerView);
		LinearLayoutManager layoutManager = new LinearLayoutManager(this);
		layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
		mRecyclerView.setLayoutManager(layoutManager);

		//Initialize Retrofit
		mService = RetrofitClient.getGithubAPIService();

		//Initialize Realm
		mRealm = Realm.getDefaultInstance();

		//Initialize NavigationView
		NavigationView navigationView = findViewById(R.id.nav_view);
		navigationView.setNavigationItemSelectedListener(this);

		//Initialize DrawerLayout
		mDrawerLayout = findViewById(R.id.drawer_layout);
		ActionBarDrawerToggle drawerToggle
				= new ActionBarDrawerToggle(this, mDrawerLayout, toolbar, R.string.drawer_open, R.string.drawer_close);
		mDrawerLayout.addDrawerListener(drawerToggle);
		drawerToggle.syncState();
	}

	//Get Intent Data and assign to variables
	private void getBundleData() {

		Intent intent = getIntent();
		if (intent.getIntExtra(Constants.KEY_QUERY_TYPE, -1) == Constants.SEARCH_BY_REPO) {
			String queryRepo = intent.getStringExtra(Constants.KEY_REPO_SEARCH);
			String repoLanguage = intent.getStringExtra(Constants.KEY_LANGUAGE);
			fetchRepositories(queryRepo, repoLanguage);
		} else {
			String githubUser = intent.getStringExtra(Constants.KEY_GITHUB_USER);
			fetchUserRepositories(githubUser);
		}
	}

	//Fetch User Repositories from Github Api
	private void fetchUserRepositories(String githubUser) {

		mService.searchRepositoriesByUser(githubUser).enqueue(new Callback<List<Repository>>() {
			@Override
			public void onResponse(@NonNull Call<List<Repository>> call, @NonNull Response<List<Repository>> response) {
				if (response.isSuccessful()) {
					Log.i(TAG, "posts loaded from API " + response);

					if(response.body() != null) {
						browsedRepositories = response.body();
					}

					if (browsedRepositories != null && browsedRepositories.size() > 0)
						setupRecyclerView(browsedRepositories);
					else
						Util.showMessage(DisplayActivity.this, "No Items Found");

				} else {
					Log.i(TAG, "Error " + response);
					if(response.errorBody() != null ) {
						Util.showErrorMessage(DisplayActivity.this, response.errorBody());
					}
				}
			}

			@Override
			public void onFailure(@NonNull Call<List<Repository>> call, @NonNull Throwable t) {
				Util.showMessage(DisplayActivity.this, t.getMessage());
				Log.e("error", t.getMessage());
				Toast.makeText(DisplayActivity.this, t.getMessage(), Toast.LENGTH_SHORT).show();
			}
		});
	}

	//Fetch Repositories from Github Api
	private void fetchRepositories(String queryRepo, String repoLanguage) {

		Map<String, String> query = new HashMap<>();

		if (repoLanguage != null && !repoLanguage.isEmpty())
			queryRepo += " language:" + repoLanguage;
		query.put("q", queryRepo);

		mService.searchRepositories(query).enqueue(new Callback<SearchResponse>() {
			@Override
			public void onResponse(@NonNull Call<SearchResponse> call, @NonNull Response<SearchResponse> response) {
				if (response.isSuccessful()) {
					Log.i(TAG, "posts loaded from API " + response);

					if(response.body() != null) {
						browsedRepositories = response.body().getItems();
					}

					if (browsedRepositories.size() > 0)
						setupRecyclerView(browsedRepositories);
					else
						Util.showMessage(DisplayActivity.this, "No Items Found");

				} else {
					Log.i(TAG, "error " + response);
					if(response.errorBody() != null ) {
						Util.showErrorMessage(DisplayActivity.this, response.errorBody());
					}

				}
			}

			@Override
			public void onFailure(@NonNull Call<SearchResponse> call, @NonNull Throwable t) {
				Util.showMessage(DisplayActivity.this, t.toString());
			}
		});
	}

	//Setup RecyclerView
	private void setupRecyclerView(List<Repository> items) {
		mDisplayAdapter = new DisplayAdapter(this, items);
		mRecyclerView.setAdapter(mDisplayAdapter);
	}

	//Setup Drawer
	@Override
	public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {

		menuItem.setChecked(true);
		closeDrawer();

		switch (menuItem.getItemId()) {

			case R.id.item_bookmark:
				showBookmarks();
				getSupportActionBar().setTitle("Showing Bookmarks");
				break;

			case R.id.item_browsed_results:
				showBrowsedResults();
				getSupportActionBar().setTitle("Showing Browsed Results");
				break;
		}

		return true;
	}

	//Swap to Repositories
	private void showBrowsedResults() {
		mDisplayAdapter.swap(browsedRepositories);
	}

	//Swap to Bookmarks
	private void showBookmarks() {
		mRealm.executeTransaction(new Realm.Transaction() {
			@Override
			public void execute(@NonNull Realm realm) {
				RealmResults<Repository> repositories = realm.where(Repository.class).findAll();
				mDisplayAdapter.swap(repositories);
			}
		});
	}

	//Close Drawer
	private void closeDrawer() {
		mDrawerLayout.closeDrawer(GravityCompat.START);
	}

	//Back Pressed
	@Override
	public void onBackPressed() {
		if (mDrawerLayout.isDrawerOpen(GravityCompat.START))
			closeDrawer();
		else {
			super.onBackPressed();
			mRealm.close();
		}
	}
}

File: activities/DisplayActivity.kt
package com.chad.gads2022_java_kotlin.activities

import android.content.Context
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.chad.gads2022_java_kotlin.R
import com.chad.gads2022_java_kotlin.adapters.DisplayAdapter
import com.chad.gads2022_java_kotlin.constants.Constants
import com.chad.gads2022_java_kotlin.extensions.showErrorMessage
import com.chad.gads2022_java_kotlin.extensions.toast
import com.chad.gads2022_java_kotlin.models.Repository
import com.chad.gads2022_java_kotlin.models.SearchResponse
import com.chad.gads2022_java_kotlin.retrofit.GithubAPIService
import com.chad.gads2022_java_kotlin.retrofit.RetrofitClient
import com.google.android.material.navigation.NavigationView
import kotlinx.android.synthetic.main.activity_display.*
import kotlinx.android.synthetic.main.header.view.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class DisplayActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {

    private lateinit var displayAdapter: DisplayAdapter
    private var repository: List<Repository> =  mutableListOf()
    private val githubAPIService: GithubAPIService by lazy {
        RetrofitClient.githubAPIService
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_display)

        //Initialize
        initialize()

        //Get Bundle Data
        bundleData()
    }

    //Initialize Views
    private fun initialize() {

        //Initialize Toolbar and set title
        setSupportActionBar(toolbar_display)
        supportActionBar!!.title = "Showing Browsed Results"

        //Initialize RecyclerView and set LayoutManager
        val layoutManager = LinearLayoutManager(this)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        recyclerView.layoutManager = layoutManager

        //Initialize NavigationView
        navigationView.setNavigationItemSelectedListener(this)

        //Initialize DrawerLayout
        val drawerToggle = ActionBarDrawerToggle(
            this,
            drawerLayout,
            toolbar_display,
            R.string.drawer_open,
            R.string.drawer_close
        )
        drawerLayout.addDrawerListener(drawerToggle)
        drawerToggle.syncState()
    }

    //Get Intent Data and assign to variables
    private fun bundleData() {
        val intent = intent
        if (intent.getIntExtra(Constants.KEY_QUERY_TYPE, -1) == Constants.SEARCH_BY_REPO) {
            val queryRepo = intent.getStringExtra(Constants.KEY_REPO_SEARCH)
            val repoLanguage = intent.getStringExtra(Constants.KEY_LANGUAGE)
            fetchRepositories(queryRepo, repoLanguage)
        } else {
            val githubUser = intent.getStringExtra(Constants.KEY_GITHUB_USER)
            fetchUserRepositories(githubUser)
        }

        setAppUserName()
    }

    private fun setAppUserName() {

        val sharedPreferences = getSharedPreferences(Constants.APP_SHARED_PREFERENCES, Context.MODE_PRIVATE)
        val name = sharedPreferences.getString(Constants.KEY_PERSON_NAME, "User")
        val headerView = navigationView.getHeaderView(0)
        headerView.txvName.text = name

    }

    //Fetch User Repositories from Github Api
    private fun fetchUserRepositories(githubUser: String?) {
        if (githubUser != null) {
            githubAPIService.searchRepositoriesByUser(githubUser).enqueue(object : Callback<List<Repository>>{
                override fun onResponse(
                    call: Call<List<Repository>>,
                    response: Response<List<Repository>>
                ) {
                    if (response.isSuccessful) {
                        Log.i(TAG, "Posts from API")

                        response.body() ?. let {
                            repository = it
                        }

                        if (repository.isNotEmpty()) {
                            setupRecyclerView(repository)
                        } else {
                            toast("No items found")
                        }

                    } else {
                        Log.i(TAG, "Error $response")
                        response.errorBody()
                            ?.let { showErrorMessage(it) }
                    }
                }

                override fun onFailure(call: Call<List<Repository>>, t: Throwable) {
                    t.message?.let { toast(it) }
                }
            })
        }
    }

    //Fetch Repositories from Github Api
    private fun fetchRepositories(queryRepo: String?, repoLanguage: String?) {
        var queryRepo = queryRepo
        val query: MutableMap<String, String?> = HashMap()
        if (repoLanguage != null && repoLanguage.isNotEmpty()) queryRepo += " language:$repoLanguage"
        query["q"] = queryRepo
        githubAPIService.searchRepositories(query).enqueue(object : Callback<SearchResponse?> {
            override fun onResponse(
                call: Call<SearchResponse?>,
                response: Response<SearchResponse?>
            ) {
                if (response.isSuccessful) {
                    Log.i(TAG, "posts loaded from API $response")

                    response.body()?.items?.let {
                        repository = it
                    }

                    if (repository.isNotEmpty()) setupRecyclerView(repository) else toast("No Items Found")
                } else {
                    Log.i(TAG, "error $response")
                    if (response.errorBody() != null) {
                        showErrorMessage(response.errorBody()!!)
                    }
                }
            }

            override fun onFailure(call: Call<SearchResponse?>, t: Throwable) {
                toast(t.toString())
            }
        })
    }

    //Setup RecyclerView
    private fun setupRecyclerView(items: List<Repository>) {
        displayAdapter = DisplayAdapter(this, items)
        recyclerView.adapter = displayAdapter
    }

    //Setup Drawer
    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        menuItem.isChecked = true
        when (menuItem.itemId) {
            R.id.item_bookmark -> {
                consumeMenuEvent({ showBookmarks() }, "Showing Bookmarks")
            }
            R.id.item_browsed_results -> {
                consumeMenuEvent({ showBrowsedResults() }, "Showing Browsed Results")
            }
        }
        return true
    }

    private fun consumeMenuEvent(function: () -> Unit, title: String) {
        function()
        closeDrawer()
        supportActionBar!!.title = title
    }

    //Swap to Repositories
    private fun showBrowsedResults() {
        displayAdapter.swap(repository)
    }

    //Swap to Bookmarks
    private fun showBookmarks() {

    }

    //Close Drawer
    private fun closeDrawer() {
        drawerLayout.closeDrawer(GravityCompat.START)
    }

    //Back Pressed
    override fun onBackPressed() {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) closeDrawer() else {
            super.onBackPressed()
        }
    }

    //Companion Object
    companion object {
        private val TAG = DisplayActivity::class.java.simpleName
    }
}